src/EfuncMsg.chpl

```diff
@@ -194,1286 +343,435 @@ module EfuncMsg
+    {
+        return wheress_helper (condition, a, b);
+    }
+
+    @arkouda.registerCommand(name="wheress_bool_float64")
+    proc wheress_br ( condition: [?d] bool, a : bool, b : real) : [d] real throws
+    {
+        return wheress_helper (condition, a, b);
+    }
+
+    @arkouda.registerCommand(name="wheress_bool_int64")
+    proc wheress_bi ( condition: [?d] bool, a : bool, b : int) : [d] int throws
+    {
+        return wheress_helper (condition, a, b);
+    }
+
+    @arkouda.registerCommand(name="wheress_bool_uint64")
+    proc wheress_bu ( condition: [?d] bool, a : bool, b : uint) : [d] uint throws
+    {
+        return wheress_helper (condition, a, b);
+    }
+
+    @arkouda.registerCommand(name="wheress_bool_bool")
+    proc wheress_bb ( condition: [?d] bool, a : bool, b : bool) : [d] bool throws
+    {
+        return wheress_helper (condition, a, b);
+    }
+
+    //  putmask has been rewritten for both the new interface and the multi-dimensional case.
+    //  The specifics are based on the observation that np.putmask behaves as if multi-dim
+    //  inputs were flattened before the operation, and then de-flattened after.
+
+   @arkouda.registerCommand(name="putmask")
+   proc putmask (mask : [?d1] bool, ref a : [d1] ?ta, v : [?d2] ?tv ) throws
+       where   (ta==real ||
+               (ta==int  && (tv==int  || tv==uint || tv==bool)) ||
+               (ta==uint && (tv==uint || tv==bool)) ||
+               (ta==bool &&  tv==bool ) ) {
+
+        // By using indexToOrder, below, this proc is generic for any rank matrices.
+
+        // Re the "de-flattened" comment above:
+
+        //    "np.putmask behaves as if multi-dim inputs were flattened before the operation."
+        //    Multi-dimensional a, mask, and v are stepped through as if one-dimensional.
+
+        //    i.e., the nth element of mask determines whether the nth element of a will be
+        //    overwritten with the nth element of v (v doesn't actually have to be the same
+        //    size as a and mask, but hold that thought for now).
+
+        //    In a departure from np.putmask, we require a and mask to have the same domain,
+        //    so that they'll be distributed identically. We don't know what that distribution
+        //    will be, and it may vary from locale to locale.
+
+        //    Further comments are embedded in the code.
+
+        // Note: orderer, copied from AryUtil.chpl, provides indexToOrder
+
+        record orderer {
+          param rank: int;
+          const accumRankSizes: [0..<rank] int;
+
+          proc init(shape: ?N*int) {
+            this.rank = N;
+            const sizesRev = [i in 0..<N] shape[N - i - 1];
+            this.accumRankSizes = * scan sizesRev / sizesRev;
+          }
+
+          inline proc indexToOrder(idx: rank*int): int {
+            var order = 0;
+            for param i in 0..<rank do order += idx[i] * accumRankSizes[rank - i - 1];
+            return order;
+          }
+
```

Comment:
Would you please move this proceedure:
```
      inline proc indexToOrder(idx: int): int { // this was added to handle the 1D case
        return idx;
      }
    }
```
to the `orderer` in `ArrayUtil` and then just use the imported order so that you are not introducing this code duplication?

Reply:
Done, and testing.  Will post follow-up comment after testing completes.

Reply:
All tests passed.  Pushing it up.

Reply:
Thanks!